src/sources/kafka.rs

```diff
@@ -370,105 +411,559 @@ impl SourceConfig for KafkaSourceConfig {
+                        }
+                    },
+
+                    ack = ack_stream.next() => match ack {
+                        Some((status, entry)) => {
+                            if status == BatchStatus::Delivered {
+                                if let Err(error) =  consumer.store_offset(&entry.topic, entry.partition, entry.offset) {
+                                    emit!(KafkaOffsetUpdateError { error });
+                                }
+                            }
+                        }
+                        None if finalizer.is_none() => {
+                            debug!("Acknowledgement stream complete for partition {}:{}.", &tp.0, tp.1);
+                            break
+                        }
+                        None => {
+                            debug!("Acknowledgement stream empty for {}:{}", &tp.0, tp.1);
+                        }
                     }
+                )
+            }
+            (tp, status)
+        });
+        (end_tx, handle)
+    }
+
+    /// Consume self, and return a "Draining" ConsumerState, along with a Future
+    /// representing a drain deadline, based on max_drain_ms
+    pub fn begin_drain(
+        self,
+        max_drain_ms: Duration,
+        sig: SyncSender<()>,
+        shutdown: bool,
+    ) -> (OptionDeadline, ConsumerStateInner<Draining>) {
+        let deadline = Box::pin(tokio::time::sleep(max_drain_ms));
+
+        let draining = ConsumerStateInner {
+            config: self.config,
+            decoder: self.decoder,
+            out: self.out,
+            log_namespace: self.log_namespace,
+            consumer_state: Draining::new(sig, shutdown),
+        };
+
+        (Some(deadline).into(), draining)
+    }
+
+    pub const fn keep_consuming(self, deadline: OptionDeadline) -> (OptionDeadline, ConsumerState) {
+        (deadline, ConsumerState::Consuming(self))
+    }
+}
+
+impl ConsumerStateInner<Draining> {
+    /// Mark the given TopicPartition as being revoked, adding it to the set of
+    /// partitions expected to drain
+    pub fn revoke_partition(&mut self, tp: TopicPartition, end_signal: oneshot::Sender<()>) {
+        // Note that if this send() returns Err, it means the task has already
+        // ended, but the completion has not been processed yet (otherwise we wouldn't have access to the end_signal),
+        // so we should still add it to the "expect to drain" set
+        _ = end_signal.send(());
+        self.consumer_state.expect_drain.insert(tp);
+    }
+
+    /// Add the given TopicPartition to the set of known "drained" partitions,
+    /// i.e. the consumer has drained the acknowledgement channel. A signal is
+    /// sent on the signal channel, indicating to the client that offsets may be committed
+    pub fn partition_drained(&mut self, tp: TopicPartition) {
+        _ = self.consumer_state.signal.send(());
```

Comment:
It might be nice to add a note about why we can safely ignore this return value.